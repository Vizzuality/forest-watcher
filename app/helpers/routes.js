// @flow

import type { Route, LocationPoint } from 'types/routes.types';
import turfBbox from '@turf/bbox';
import { type BBox2d, lineString as turfLineString } from '@turf/helpers';

export function bbox(route: Route): BBox2d {
  const routePoints = [...route.locations, route.destination];
  const lineString = turfLineString(
    routePoints.map((item: Location | LocationPoint) => [item.longitude, item.latitude])
  );
  return turfBbox(lineString);
}
