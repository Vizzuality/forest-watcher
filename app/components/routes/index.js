// @flow

import React, { PureComponent } from 'react';
import { View, Text, ScrollView, Platform } from 'react-native';

import moment from 'moment';
import i18n from 'i18next';
import debounceUI from 'helpers/debounceUI';
import tracker from 'helpers/googleAnalytics';
import styles from './styles';
import { Navigation } from 'react-native-navigation';

import EmptyState from 'components/common/empty-state';
import RoutePath from 'components/common/route-path';

import type { Route } from 'types/routes.types';

import ShareSheet from 'components/common/share';
import VerticalSplitRow from 'components/common/vertical-split-row';

import { formatDistance, getDistanceOfPolyline } from 'helpers/map';
import { isSmallScreen } from 'config/theme';

import exportLayerManifest from 'helpers/sharing/exportLayerManifest';
import manifestBundleSize from 'helpers/sharing/manifestBundleSize';
import generateUniqueID from 'helpers/uniqueId';
import { formatBytes } from 'helpers/data';

const nextIcon = require('assets/next.png');
const emptyIcon = require('assets/routesEmpty.png');
const routeMapBackground = require('assets/routeMapBackground.png');

const RoutePreviewSize = isSmallScreen ? 86 : 122;

type Props = {|
  +componentId: string,
  +exportRoutes: (ids: Array<string>) => Promise<void>,
  +routes: Array<Route>,
  +initialiseAreaLayerSettings: (string, string) => void,
  +setSelectedAreaId: (areaId: string) => void,
  +showExportReportsSuccessfulNotification: () => void
|};

type State = {|
  +bundleSize: number | typeof undefined,
  +selectedForExport: Array<string>,
  +inShareMode: boolean
|};

export default class Routes extends PureComponent<Props, State> {
  static options(passProps: {}) {
    return {
      topBar: {
        title: {
          text: i18n.t('dashboard.routes')
        }
      }
    };
  }

  shareSheet: any;

  constructor(props: Props) {
    super(props);

    // Set an empty starting state for this object. If empty, we're not in export mode. If there's items in here, export mode is active.
    this.state = {
      inShareMode: false,
      selectedForExport: []
    };
  }

  componentDidMount() {
    tracker.trackScreenView('My Routes');
  }

  fetchExportSize = async (routeIds: Array<string>) => {
    const currentFetchId = generateUniqueID();
    this.fetchId = currentFetchId;
    this.setState({
      bundleSize: undefined
    });
    const manifest = await exportLayerManifest(
      {
        areaIds: [],
        basemapIds: [],
        layerIds: [],
        reportIds: [],
        routeIds
      },
      [],
      this.props.routes.filter(route => {
        return routeIds.includes(route.id);
      }),
      []
    );
    const fileSize = manifestBundleSize(manifest);
    if (this.fetchId == currentFetchId) {
      this.setState({
        bundleSize: fileSize
      });
    }
  };

  onFrequentlyAskedQuestionsPress = () => {
    Navigation.push(this.props.componentId, {
      component: {
        name: 'ForestWatcher.FaqCategories'
      }
    });
  };

  /**
   * Handles the route row being selected while in export mode.
   * Will swap the state for the specified row, to show in the UI if it has been selected or not.
   */
  onRouteSelectedForExport = (route: Route) => {
    this.setState(state => {
      if (state.selectedForExport.includes(route.areaId + route.id)) {
        const selectedForExport = [...state.selectedForExport].filter(id => route.areaId + route.id != id);
        this.fetchExportSize(selectedForExport);
        return {
          selectedForExport
        };
      } else {
        const selected = [...state.selectedForExport];
        selected.push(route.areaId + route.id);
        this.fetchExportSize(selected);
        return {
          selectedForExport: selected
        };
      }
    });
  };

  /**
   * Handles a route row being selected.
   */
  onClickRoute = debounceUI((route: Route) => {
    // Testing against a mocked route? You must provide your own area id here!
    this.props.setSelectedAreaId(route.areaId);
    this.props.initialiseAreaLayerSettings(route.id, route.areaId);
    Navigation.push(this.props.componentId, {
      component: {
        name: 'ForestWatcher.Map',
        options: {
          topBar: {
            title: {
              text: route.name
            }
          }
        },
        passProps: {
          previousRoute: route
        }
      }
    });
  });

  onClickRouteSettings = debounceUI((route: Route) => {
    Navigation.showModal({
      stack: {
        children: [
          {
            component: {
              name: 'ForestWatcher.RouteDetail',
              passProps: {
                routeId: route.id,
                routeName: route.name
              }
            }
          }
        ]
      }
    });
  });

  /**
   * Handles the 'export <x> routes' button being tapped.
   *
   * @param  {Object} selectedRoutes A mapping of route identifiers to a boolean dictating whether they've been selected for export.
   */
  onExportRoutesTapped = debounceUI(selectedRoutes => {
    // TODO: Loading screen while the async function below executed
    this.props.exportRoutes(selectedRoutes);
    this.shareSheet?.setSharing?.(false);
    this.setSharing(false);
  });

  setAllSelected = (selected: boolean) => {
    const selectedForExport = selected ? this.props.routes.map(route => route.areaId + route.id) : [];
    this.fetchExportSize(selectedForExport);
    this.setState({
      selectedForExport
    });
  };

  setSharing = (sharing: boolean) => {
    // Can set selectedForExport to [] either way as we want to start sharing again with none selected
    this.setState({
      bundleSize: undefined,
      inShareMode: sharing,
      selectedForExport: []
    });
  };

  /**
   * Sorts oroutes in descending date order
   *
   * @param {Array} routes <Route> An array of routes
   * @return {Array} The routes, but sorted...
   */
  sortedRoutes(routes: Array<Route>) {
    const sorted = [...routes];
    sorted.sort((a, b) => {
      if (a.date > b.date) {
        return -1;
      }
      if (a.date < b.date) {
        return +1;
      }
      return 0;
    });
    return sorted;
  }

  /**
   * Renders a component for the route's path to be shown on the row
   * @param <Route> route The route to render a path for
   */
  renderRoutePath = (route: Route) => {
    return <RoutePath size={RoutePreviewSize} route={route} />;
  };

  /**
   * renderItems - Returns an array of rows, based on the route data provided.
   *
   * @param  {Array} data <Route>  An array of routes.
   * @param  {any} image                The action image.
   * @param  {void} onPress             The action callback.
   * @return {Array}                    An array of route rows.
   */
  renderItems(data: Array<Route>, image: any, onPress: string => void) {
    return this.sortedRoutes(data).map((item, index) => {
      const routeDistance = getDistanceOfPolyline(item.locations);
      const dateText = moment(item.endDate).format('ll');
      const distanceText = formatDistance(routeDistance, 1, false);
      const subtitle = dateText + ', ' + distanceText;

      const combinedId = item.areaId + item.id;

      return (
        <VerticalSplitRow
          backgroundImageResizeMode={Platform.OS === 'ios' ? 'repeat' : 'cover'}
          key={combinedId}
          onSettingsPress={this.onClickRouteSettings.bind(this, item)}
          onPress={() => {
            onPress(item);
          }}
          style={styles.row}
          renderImageChildren={this.renderRoutePath.bind(this, item)}
          imageSrc={routeMapBackground}
          title={item.name}
          subtitle={subtitle}
          disableSettingsButton={this.state.inShareMode}
          selected={this.state.inShareMode ? this.state.selectedForExport.includes(combinedId) : null}
          largerLeftPadding
          largeImage
        />
      );
    });
  }

  renderSection(title: string, ...options: [Array<Route>, any, (string) => void]) {
    return (
      <View style={styles.listContainer}>
        <View style={styles.listHeader}>
          <Text style={styles.listTitle}>{title}</Text>
        </View>
        {this.renderItems(...options)}
      </View>
    );
  }

  /**
   * renderRoutes - Renders a list of routes.
   *
   * @param  {array} routes      An array of routes.
   * @param  {bool} inExportMode  Whether the user is in export mode or not. If in export mode, a different callback will be used.
   * @return {ScrollView}         A ScrollView element with all content rendered to it.
   */
  renderRoutes(routes: Array<Route>, inExportMode: boolean) {
    const hasRoutes = !!routes.length;

    if (!hasRoutes) {
      return (
        <View style={styles.containerEmpty}>
          <EmptyState
            actionTitle={i18n.t('routes.empty.action')}
            body={i18n.t('routes.empty.body')}
            onActionPress={this.onFrequentlyAskedQuestionsPress}
            icon={emptyIcon}
            title={i18n.t('routes.empty.title')}
          />
        </View>
      );
    }

    return (
      <ScrollView
        style={styles.container}
        contentContainerStyle={styles.contentContainer}
        showsVerticalScrollIndicator={false}
        showsHorizontalScrollIndicator={false}
      >
        <View style={styles.container}>
          {routes &&
            routes.length > 0 &&
            this.renderSection(
              i18n.t('routes.myRoutes'),
              routes,
              nextIcon,
              inExportMode ? this.onRouteSelectedForExport : this.onClickRoute
            )}
        </View>
      </ScrollView>
    );
  }

  /**
   * getShareButtonText - given the total number of routes to share, returns the
   * text that should be shown in the button.
   *
   * @param {number} totalToShare the amount of routes that should be shared.
   * @param {?number} bundleSize the size of the shareable bundle.
   *
   * @returns {string}
   */
  getShareButtonText = (totalToShare: number, bundleSize: ?number): string => {
    if (totalToShare === 0) {
      return i18n.t('routes.sharing.noneSelected');
    }

    let transifexKey = 'routes.sharing.multipleRoutes';

    if (totalToShare === 1) {
      transifexKey = 'routes.sharing.oneRoute';
    }

    return i18n.t(transifexKey, {
      bundleSize: bundleSize !== undefined ? formatBytes(bundleSize) : i18n.t('commonText.calculating')
    });
  };

  render() {
    // Determine if we're in export mode, and how many routes have been selected to export.
    const totalToExport = this.state.selectedForExport.length;
    const totalRoutes = this.props.routes.length;

    return (
      /* View necessary to fix the swipe back on wix navigation */
      <View style={styles.container}>
        <ShareSheet
          componentId={this.props.componentId}
          enabled={totalToExport > 0}
          onShare={() => {
            this.onExportRoutesTapped(this.state.selectedForExport);
          }}
          onSharingToggled={this.setSharing}
          onToggleAllSelected={this.setAllSelected}
          ref={ref => {
            this.shareSheet = ref;
          }}
          selected={totalToExport}
          selectAllCountText={
            totalRoutes > 1
              ? i18n.t('routes.export.manyRoutes', { count: totalRoutes })
              : i18n.t('routes.export.oneRoute', { count: 1 })
          }
          shareButtonDisabledTitle={i18n.t('routes.sharing.title')}
          shareButtonEnabledTitle={this.getShareButtonText(totalToExport, this.state.bundleSize)}
        >
          {this.renderRoutes(this.props.routes, this.state.inShareMode)}
        </ShareSheet>
      </View>
    );
  }
}
