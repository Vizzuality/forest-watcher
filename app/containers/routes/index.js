// @flow
import type { State } from 'types/store.types';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { setSelectedAreaId } from 'redux-modules/areas';
import { shouldBeConnected } from 'helpers/app';

import { showExportReportsSuccessfulNotification } from 'redux-modules/app';

import Routes from 'components/routes';
import { initialiseAreaLayerSettings } from 'redux-modules/layerSettings';

function mapStateToProps(state: State) {
  return {
    appLanguage: state.app.language,
    isConnected: shouldBeConnected(state),
    routes: state.routes.previousRoutes
  };
}

function mapDispatchToProps(dispatch: *) {
  return bindActionCreators(
    {
      initialiseAreaLayerSettings,
      setSelectedAreaId,
      showExportReportsSuccessfulNotification
    },
    dispatch
  );
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Routes);
