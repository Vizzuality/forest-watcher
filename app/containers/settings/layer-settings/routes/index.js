// @flow
import type { State } from 'types/store.types';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { DEFAULT_LAYER_SETTINGS, deselectAllRoutes, toggleRouteSelected } from 'redux-modules/layerSettings';
import RoutesLayerSettings from 'components/settings/layer-settings/routes';
import { getAllRouteIds } from 'redux-modules/routes';

function mapStateToProps(state: State, ownProps) {
  return {
    featureId: ownProps.featureId,
    myRoutes: state.routes.previousRoutes,
    importedRoutes: state.routes.importedRoutes,
    routesLayerSettings: state.layerSettings?.[ownProps.featureId]?.routes || DEFAULT_LAYER_SETTINGS.routes
  };
}

const mapDispatchToProps = (dispatch: *) => {
  return {
    ...bindActionCreators(
      {
        deselectAllRoutes
      },
      dispatch
    ),
    toggleRouteSelected: (featureId: string, routeId: string) => {
      const allRouteIds = dispatch(getAllRouteIds());
      dispatch(toggleRouteSelected(featureId, routeId, allRouteIds));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RoutesLayerSettings);
