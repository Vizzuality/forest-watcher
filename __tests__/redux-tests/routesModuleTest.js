// @flow
import 'react-native';
import { migrateV1RoutesToV2RoutesStructure } from 'redux-modules/routes';

function getMockUUID() {
  return `487a2312-7600-4c9f-a7ff-d6ad69a721a0`;
}

describe('redux-modules/routes', () => {
  describe('func migrateV1RoutesToV2RoutesStructure', () => {
    const mockArea1 = {
      name: 'testArea1',
      id: 'testAreaId1',
      application: 'applicationMock', // used to test that all fields are included in payload
      geostore: { id: 'testGeostoreId1' }
    };

    const mockV1Route1 = {
      id: 'testRouteId1',
      areaId: 'testAreaId1',
      name: 'routeNameMock'
    };

    const mockV1Route2 = {
      id: 'testRouteId2',
      areaId: 'testAreaId1',
      name: 'routeNameMock'
    };

    const mockV2Route = {
      id: '487a2312-7600-4c9f-a7ff-d6ad69a721a0',
      areaId: 'testAreaId1',
      name: 'routeNameMock',
      geostoreId: 'testGeostoreId1'
    };

    const mockV1RouteStateSingleRoute = {
      activeRoute: undefined,
      previousRoutes: [mockV1Route1]
    };

    const mockV1RouteStateActiveRoute = {
      activeRoute: {
        areaId: 'testAreaId1',
        id: '1234567890',
        startDate: 1234567890
      },
      previousRoutes: []
    };

    const mockV1RouteStateMultipleRoutes = {
      activeRoute: undefined,
      previousRoutes: [mockV1Route1, mockV1Route2]
    };

    const mockV1RouteStateMultipleRoutesForDifferentAreas = {
      activeRoute: undefined,
      previousRoutes: [mockV1Route1, { ...mockV1Route2, areaId: 'testAreaId2' }]
    };

    it('returns fresh state when provided no params', () => {
      expect(migrateV1RoutesToV2RoutesStructure()).toMatchSnapshot();
    });

    it('returns fresh state when provided null params', () => {
      expect(migrateV1RoutesToV2RoutesStructure(null, null)).toMatchSnapshot();
    });

    it('returns fresh state when provided undefined params', () => {
      expect(migrateV1RoutesToV2RoutesStructure(undefined)).toMatchSnapshot();
    });

    it('returns fresh state when provided invalid params', () => {
      expect(migrateV1RoutesToV2RoutesStructure(mockV1Route1, mockArea1, getMockUUID)).toMatchSnapshot();
    });

    it('returns updated state when provided route state with no routes', () => {
      expect(
        migrateV1RoutesToV2RoutesStructure(
          { ...mockV1RouteStateSingleRoute, previousRoutes: [] },
          [mockArea1],
          getMockUUID
        )
      ).toMatchSnapshot();
    });

    it('returns existing state when provided route state already using v2 structure', () => {
      expect(
        migrateV1RoutesToV2RoutesStructure(
          { ...mockV1RouteStateSingleRoute, previousRoutes: [mockV2Route], routeStructureVersion: 'v2' },
          [mockArea1],
          getMockUUID
        )
      ).toMatchSnapshot();
    });

    describe('with active route', () => {
      it('returns updated state when provided null areas', () => {
        expect(migrateV1RoutesToV2RoutesStructure(mockV1RouteStateActiveRoute, null, getMockUUID)).toMatchSnapshot();
      });

      it('returns updated state when provided empty areas', () => {
        expect(migrateV1RoutesToV2RoutesStructure(mockV1RouteStateActiveRoute, [], getMockUUID)).toMatchSnapshot();
      });

      it('returns updated state when provided area with non-matching ID', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateActiveRoute,
            [{ ...mockArea1, id: 'nonMatchingID' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });

      it('returns updated state when provided areas with non-matching ID', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateActiveRoute,
            [{ ...mockArea1, id: 'nonMatchingID1' }, { ...mockArea1, id: 'nonMatchingID2' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });

      it('returns updated state when provided area with matching ID', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(mockV1RouteStateActiveRoute, [mockArea1], getMockUUID)
        ).toMatchSnapshot();
      });

      it('returns updated state when provided areas where one area matches', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateActiveRoute,
            [mockArea1, { ...mockArea1, id: 'nonMatchingId' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });
    });

    describe('with one route', () => {
      it('returns updated state when provided null areas', () => {
        expect(migrateV1RoutesToV2RoutesStructure(mockV1RouteStateSingleRoute, null, getMockUUID)).toMatchSnapshot();
      });

      it('returns updated state when provided empty areas', () => {
        expect(migrateV1RoutesToV2RoutesStructure(mockV1RouteStateSingleRoute, [], getMockUUID)).toMatchSnapshot();
      });

      it('returns updated state when provided area with non-matching ID', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateSingleRoute,
            [{ ...mockArea1, id: 'nonMatchingID' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });

      it('returns updated state when provided areas with non-matching ID', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateSingleRoute,
            [{ ...mockArea1, id: 'nonMatchingID1' }, { ...mockArea1, id: 'nonMatchingID2' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });

      it('returns updated state when provided area with matching ID', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(mockV1RouteStateSingleRoute, [mockArea1], getMockUUID)
        ).toMatchSnapshot();
      });

      it('returns updated state when provided areas where one area matches', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateSingleRoute,
            [mockArea1, { ...mockArea1, id: 'nonMatchingId' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });
    });

    describe('with two routes', () => {
      it('returns updated state when provided null areas', () => {
        expect(migrateV1RoutesToV2RoutesStructure(mockV1RouteStateMultipleRoutes, null, getMockUUID)).toMatchSnapshot();
      });

      it('returns updated state when provided empty areas', () => {
        expect(migrateV1RoutesToV2RoutesStructure(mockV1RouteStateMultipleRoutes, [], getMockUUID)).toMatchSnapshot();
      });

      it('returns updated state when provided area with non-matching ID', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateMultipleRoutes,
            [{ ...mockArea1, id: 'nonMatchingID' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });

      it('returns updated state when provided areas with non-matching ID', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateMultipleRoutes,
            [{ ...mockArea1, id: 'nonMatchingID1' }, { ...mockArea1, id: 'nonMatchingID2' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });

      it('returns updated state when provided area with matching ID', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(mockV1RouteStateMultipleRoutes, [mockArea1], getMockUUID)
        ).toMatchSnapshot();
      });

      it('returns updated state when provided areas where one area matches', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateMultipleRoutes,
            [mockArea1, { ...mockArea1, id: 'nonMatchingId' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });

      it('returns updated state when provided routes for different areas & neither match', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateMultipleRoutesForDifferentAreas,
            [{ ...mockArea1, id: 'nonMatchingId1' }, { ...mockArea1, id: 'nonMatchingId2' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });

      it('returns updated state when provided routes for different areas & only one matches', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateMultipleRoutesForDifferentAreas,
            [mockArea1, { ...mockArea1, id: 'nonMatchingId' }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });

      it('returns updated state when provided routes for different areas & both match', () => {
        expect(
          migrateV1RoutesToV2RoutesStructure(
            mockV1RouteStateMultipleRoutesForDifferentAreas,
            [mockArea1, { ...mockArea1, id: 'testAreaId2', geostore: { id: 'testGeostoreId2' } }],
            getMockUUID
          )
        ).toMatchSnapshot();
      });
    });
  });
});
