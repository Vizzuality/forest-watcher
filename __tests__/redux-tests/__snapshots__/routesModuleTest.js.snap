// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 returns existing state when provided route state already using v2 structure 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": "testGeostoreId1",
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
  "routeStructureVersion": "v2",
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 returns fresh state when provided invalid params 1`] = `undefined`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 returns fresh state when provided no params 1`] = `undefined`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 returns fresh state when provided null params 1`] = `undefined`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 returns fresh state when provided undefined params 1`] = `undefined`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 returns updated state when provided route state with no routes 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with active route returns updated state when provided area with matching ID 1`] = `
Object {
  "activeRoute": Object {
    "areaId": "testAreaId1",
    "geostoreId": "testGeostoreId1",
    "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
    "startDate": 1234567890,
  },
  "previousRoutes": Array [],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with active route returns updated state when provided area with non-matching ID 1`] = `
Object {
  "activeRoute": Object {
    "areaId": "testAreaId1",
    "geostoreId": undefined,
    "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
    "startDate": 1234567890,
  },
  "previousRoutes": Array [],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with active route returns updated state when provided areas where one area matches 1`] = `
Object {
  "activeRoute": Object {
    "areaId": "testAreaId1",
    "geostoreId": "testGeostoreId1",
    "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
    "startDate": 1234567890,
  },
  "previousRoutes": Array [],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with active route returns updated state when provided areas with non-matching ID 1`] = `
Object {
  "activeRoute": Object {
    "areaId": "testAreaId1",
    "geostoreId": undefined,
    "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
    "startDate": 1234567890,
  },
  "previousRoutes": Array [],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with active route returns updated state when provided empty areas 1`] = `
Object {
  "activeRoute": Object {
    "areaId": "testAreaId1",
    "geostoreId": undefined,
    "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
    "startDate": 1234567890,
  },
  "previousRoutes": Array [],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with active route returns updated state when provided null areas 1`] = `
Object {
  "activeRoute": Object {
    "areaId": "testAreaId1",
    "geostoreId": undefined,
    "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
    "startDate": 1234567890,
  },
  "previousRoutes": Array [],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with one route returns updated state when provided area with matching ID 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": "testGeostoreId1",
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with one route returns updated state when provided area with non-matching ID 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with one route returns updated state when provided areas where one area matches 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": "testGeostoreId1",
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with one route returns updated state when provided areas with non-matching ID 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with one route returns updated state when provided empty areas 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with one route returns updated state when provided null areas 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with two routes returns updated state when provided area with matching ID 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": "testGeostoreId1",
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
    Object {
      "areaId": "testAreaId1",
      "geostoreId": "testGeostoreId1",
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with two routes returns updated state when provided area with non-matching ID 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with two routes returns updated state when provided areas where one area matches 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": "testGeostoreId1",
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
    Object {
      "areaId": "testAreaId1",
      "geostoreId": "testGeostoreId1",
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with two routes returns updated state when provided areas with non-matching ID 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with two routes returns updated state when provided empty areas 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with two routes returns updated state when provided null areas 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with two routes returns updated state when provided routes for different areas & both match 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": "testGeostoreId1",
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
    Object {
      "areaId": "testAreaId2",
      "geostoreId": "testGeostoreId2",
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with two routes returns updated state when provided routes for different areas & neither match 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
    Object {
      "areaId": "testAreaId2",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;

exports[`redux-modules/routes func migrateRouteStateFromV1ToV2 with two routes returns updated state when provided routes for different areas & only one matches 1`] = `
Object {
  "activeRoute": undefined,
  "previousRoutes": Array [
    Object {
      "areaId": "testAreaId1",
      "geostoreId": "testGeostoreId1",
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
    Object {
      "areaId": "testAreaId2",
      "geostoreId": undefined,
      "id": "487a2312-7600-4c9f-a7ff-d6ad69a721a0",
      "name": "routeNameMock",
    },
  ],
}
`;
